File: streaming/flutter_app/lib/models/match.dart
import 'fighter.dart';
import 'odds.dart';

enum MatchStatus { upcoming, live, completed, cancelled }

class Match {
  final String id;
  final Fighter fighter1;
  final Fighter fighter2;
  final MatchStatus status;
  final String? streamUrl;
  final DateTime scheduledAt;
  final DateTime? completedAt;
  final String? winnerId;
  final double totalPool;
  final int activeBets;
  final Odds odds;
  final String label;
  final int bestOf;
  final int currentRound;
  final int roundsWonP1;
  final int roundsWonP2;
  final bool bettingOpen;

  const Match({
    required this.id,
    required this.fighter1,
    required this.fighter2,
    required this.status,
    this.streamUrl,
    required this.scheduledAt,
    this.completedAt,
    this.winnerId,
    this.totalPool = 0,
    this.activeBets = 0,
    required this.odds,
    this.label = 'MK4-Classic',
    this.bestOf = 3,
    this.currentRound = 1,
    this.roundsWonP1 = 0,
    this.roundsWonP2 = 0,
    this.bettingOpen = false,
  });

  factory Match.fromJson(Map<String, dynamic> json) {
    final oddsJson = json['odds'] as Map<String, dynamic>?;
    final odds = oddsJson != null ? Odds.fromJson(oddsJson) : Odds.zero;

    return Match(
      id: json['id'] as String,
      fighter1: Fighter.fromJson(json['fighter1'] as Map<String, dynamic>),
      fighter2: Fighter.fromJson(json['fighter2'] as Map<String, dynamic>),
      status: _parseStatus(json['status'] as String),
      streamUrl: json['stream_url'] as String?,
      scheduledAt: DateTime.parse(json['scheduled_at'] as String),
      completedAt: json['completed_at'] != null
          ? DateTime.parse(json['completed_at'] as String)
          : null,
      winnerId: json['winner_id'] as String?,
      totalPool: (odds.fighter1PoolPct + odds.fighter2PoolPct) > 0
          ? (oddsJson?['total_pool'] as num?)?.toDouble() ?? 0.0
          : 0.0,
      activeBets: (oddsJson?['active_bets'] as int?) ?? 0,
      odds: odds,
      label: json['label'] as String? ?? 'MK4-Classic',
      bestOf: json['best_of'] as int? ?? 3,
      currentRound: json['current_round'] as int? ?? 1,
      roundsWonP1: json['rounds_won_p1'] as int? ?? 0,
      roundsWonP2: json['rounds_won_p2'] as int? ?? 0,
      bettingOpen: json['betting_open'] as bool? ?? false,
    );
  }

  static MatchStatus _parseStatus(String s) {
    switch (s) {
      case 'live':
        return MatchStatus.live;
      case 'completed':
        return MatchStatus.completed;
      case 'cancelled':
        return MatchStatus.cancelled;
      default:
        return MatchStatus.upcoming;
    }
  }
}

File: backend/app/schemas/match.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel

from app.schemas.fighter import FighterOut


class OddsOut(BaseModel):
    fighter1_odds: float
    fighter2_odds: float
    fighter1_pool_pct: float
    fighter2_pool_pct: float
    fighter1_pool: float = 0.0
    fighter2_pool: float = 0.0
    total_pool: float
    active_bets: int


class MatchBetFeedOut(BaseModel):
    wallet_masked: str
    side: str
    fighter_name: str
    amount: float
    status: str
    placed_at: datetime


class MatchOut(BaseModel):
    id: UUID
    fighter1: FighterOut | None = None
    fighter2: FighterOut | None = None
    status: str
    label: str
    scheduled_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None
    winner_id: UUID | None = None
    stream_url: str | None = None
    odds: OddsOut | None = None
    best_of: int = 3
    current_round: int = 1
    rounds_won_p1: int = 0
    rounds_won_p2: int = 0
    betting_open: bool = False
    queue_position: int | None = None
    queue_starts_at: datetime | None = None
    queue_countdown_seconds: int | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class MatchCreate(BaseModel):
    fighter1_id: UUID
    fighter2_id: UUID
    scheduled_at: datetime
    label: str = "MK4-Classic"
    savestate_path: str | None = None
    p1_agent: str = "random"
    p2_agent: str = "random"
    best_of: int = 3
